# 异常
> 程序执行过程中出现的非正常的情况，最终会导致JVM的非正常停止。<br>
 异常本身是一个`类`，产生异常就是创建异常对象并抛出了一个`异常对象`。<br>
## 异常体系
### Throwable
* Throwable类是Java语言中所有错误或异常的超类，即父类，包括两个子类：Error，Exception。
#### Error
* 错误就相当于无法治愈的毛病，必须修改代码
#### `Exception`
>       编译期异常，进行编译java程序出现问题
>       RuntimeException:运行期异常，java程序运行出现问题，如数组越界异常
## 异常产生过程解析（分析与处理）
### 以数组越界异常为例：~~插入图片~~
## 异常的处理
* 处理异常的关键字：try，catch，finally，throw，throws
### 抛出异常throw关键字
#### 作用
可以使用throw关键字在指定的方法中抛出指定的异常
#### 使用格式
throw new xxxException(“异常产生的原因”）
#### 注意
>       1、throw关键字必须写在方法的内部
>       2、throw关键字后面new的对象必须是Exception或者Exception的子类对象
>       3、throw关键字抛出指定的异常对象，我们就必须处理这个异常对象
>>      throw关键字后边创建的是RuntimeException或者是RuntimeException的子类对象，我们可以不处理，默认交给JVM处理（打印异常对象，中断程序）
>>      throw关键字后面创建的是编译异常（写代码的时候报错），我们就必须处理这个异常，要么throws，要么try...catch
>       以后我们首先必须对方法传递过来的参数进行合法性校验，如果参数不合法，就必须抛出异常，告知调用者
#### 例子
* 如方法传递进来的数组为null，则可以抛出异常
>>      throw new NullPointerException("传递的数组值是null");
>>      空指针异常实际上是运行期异常，默认交给JVM处理，我们可不处理
* 数组越界异常的处理
>>      throw new ArrayIndexOutOfBoundsException（“数组越界”）；
* 判断对象是否为空
>>      Objects.requireNonNull(obj,"信息");//会打印异常
### throws关键字：异常处理的第一种方式，交给别人处理
#### 作用
>       当方法内部抛出异常对象的时候，那么就必须处理
>       可以使用throws关键字处理异常对象，会把异常对象生命抛出给方法的调用者处理，最终交给JVM处理---->中断处理
#### 使用格式：在方法声明时使用
>       修饰符 返回值类型 方法名（参数列表） throws AAAException, BBBException...{
>           throw new AAAException("产生原因");
>           throw new BBBException("产生原因");
>           ...
>       }
>>      注意：
>>          1、throws关键字必须卸载方法声明处
>>          2、throws关键字后边声明的异常必须是Exception或者其子类
>>          3、若多个对象有子父类关系，那么直接声明父类异常既可
* 例子
> 定义一个方法，对传递的文件路径进行合法性判断，使用方法时要`声明这个异常`
>>      throw new FileNotFoundException("文件路径不对 ");
### try...catch方法
* 按上述方法，在抛出异常后不会执行后面的代码，这时就要用到try...catch
#### 格式
```      
try{
    可能产生异常的代码
}catch(定义一个异常的变量，用来接收try中抛出的异常对象){
    异常的处理逻辑，抛出异常对象之后，怎么处理异常对象
    一般在工作中，会把异常的信息记录到一个日志中，可以在这输出一句话
}
...
catch（异常类名 变量名）{

}
```
#### finally代码块
```      
try{
    可能产生异常的代码
}catch(定义一个异常的变量，用来接收try中抛出的异常对象){
    异常的处理逻辑，抛出异常对象之后，怎么处理异常对象
    一般在工作中，会把异常的信息记录到一个日志中，可以在这输出一句话
}
...
catch（异常类名 变量名）{

}finally{
    无论是否出现异常都会执行
}
```
* 注意
>       1、try中可能会抛出多个异常对象，那么就可以使用多个catch来处理这些异常对象
>       2、如果try中产生了异常，那么就会执行catch中的异常处理逻辑，之后继续执行try...catch之后代码
>          如果try中没有异常，那么就`不`会执行catch中的异常处理逻辑，之后继续执行try...catch之后代码
>       3、finally不能单独使用，必须和try一起使用
>       4、finally一般用于资源释放与回收，无论程序是否出现异常，最后都要回收资源（IO流）
>       5、如果finally有return语句，则必返回finally中的结果，要避免。
### Throwable类中定义了3个异常处理的方法
>       String getMessage()返回此Throwable的简短描述
>       String toString()返回此Throwable的详细消息字符串
>       void printStackTrace（） JVM打印异常对象，默认此方法，打印的异常信息最全面
### 多个异常处理办法（try...catch）
>       1、多次捕获多次处理，即用多个try...catch
>       2、一次捕获多次处理，即用一个try多个catch
>>      2的注意事项：catch里边定义的异常变量，如果有子父类关系，那么子类的异常变量必须写在上边，否则就会报错
>>                  如ArrayIndexOutOfBoundsException（数组越界异常） extends IndexOutOfBoundsException（集合越界异常）
>>                  否则try抛出的异常对象都会给父类，定义的子类参数就没用，会报错
>       3、多次捕获多次处理，即用一个try...catch，要求catch中的异常类型为try中所有异常的父类，就可以实现多态。
### 子父类异常处理：父类异常什么样，子类就什么样。
>       若父类抛出多个异常，子类重写父类方法时，可抛出相同或父类异常子类或不抛异常
>       父类方法没有抛出异常，子类重写父类方法时可不抛。若子类产生异常，只能捕获处理（try...catch），不能声明抛出。
## 自定义异常
### 格式
>       public class XXXException extends Exception / RuntimeException{
>           添加一个空参数的构造方法
>           添加一个带异常信息的构造方法
>       }
>>      注意：
>>          1、自定义异常类一般都是以Exception结尾，说明该类是一个异常类
>>          2、自定义异常类，必须的继承Exception或者RuntimeException
>>              继承Exception：那么自定义的异常类就是一个编译器异常，如果方法内部抛出了编译期异常，就必须处理这个异常，要么throws，要么try...catch，不处理就会报错。
>>              继承RuntimeException：那么自定义的异常类就是一个运行期异常，交给虚拟机处理（中断处理），意思就是可以不throws声明和try...catch等也不会报错。

## 总结：运行时异常，可交给虚拟机处理，编译时异常则必须处理。
## 两个概念：抛出异常（throw）和处理异常（throws和try...catch等方法）














